crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -0,0 +1,720 @@
+use super::node::{Instruction, Operation};
+use acvm::FieldElement;
+use arena::Index;
+use num_traits::ToPrimitive;
+use std::collections::HashMap;
+//use crate::acir::native_types::{Arithmetic, Witness};
+use crate::ssa::code_gen::IRGenerator;
+use crate::ssa::node;
+use crate::ssa::node::Node;
+use crate::Evaluator;
+use crate::Gate;
+use crate::RuntimeErrorKind;
+use acvm::acir::circuit::gate::Directive;
+use acvm::acir::native_types::{Arithmetic, Linear, Witness};
+use num_bigint::BigUint;
+use std::convert::TryInto;
+pub struct Acir {
+    pub arith_cache: HashMap<Index, InternalVar>,
+}
+
+#[derive(Clone, Debug)]
+pub struct InternalVar {
+    expression: Arithmetic,
+    //value: FieldElement,     //not used for now
+    witness: Option<Witness>,
+    idx: Option<Index>,
+}
+impl InternalVar {
+    pub fn is_equal(&self, b: &InternalVar) -> bool {
+        if self.idx.is_some() && b.idx.is_some() && self.idx.unwrap() == b.idx.unwrap() {
+            return true;
+        }
+        if self.witness.is_some()
+            && b.witness.is_some()
+            && self.witness.unwrap() == b.witness.unwrap()
+        {
+            return true;
+        }
+        false //TODO should we check if content is the same??
+    }
+}
+
+impl Acir {
+    //This function stores the substitution with the arithmetic expression in the cache
+    //When an instruction performs arithmetic operation, its output can be represented as an arithmetic expression of its arguments
+    //Substitute a nodeobj as an arithmetic expression
+    fn substitute(
+        arith_cache: &mut HashMap<Index, InternalVar>,
```

Comment:
It could be, however I wonder if it is not better to tell rust what will be modified exactly rather than moving the whole 'self'?

Reply:
I think the approach with self is slightly better since self only contains the one hashmap anyways. It gives you some extra sugar so you can write
```rs
let l_c = self.substitute(ins.lhs, evaluator, cfg);
```
instead of
```rs
let l_c = Acir::substitute(&mut self.arith_cache, ins.lhs, evaluator, cfg);
```

Reply:
Done